replication/binlogsyncer.go

```diff
@@ -382,6 +401,12 @@ func (b *BinlogSyncer) enableSemiSync() error {
 	return nil
 }
 
+func (b *BinlogSyncer) SetEventHandler(handler EventHandler) {
```

Comment:
The change to require a handler be defined on `BinlogSyncer` instantiation would break the following flow:
```
cfg := BinlogSyncerConfig{...}
syncer := NewBinlogSyncer(cfg)
syncer.StartBackup(...)
```
where `StartBackup` calls `StartBackupWithHandler` which instantiates the handler. It'd instead have to be something like:
```
cfg := BinlogSyncerConfig{EventHandleMode: EventHandleModeSync}
handler := BackupEventHandler{handler: func(filename string) (io.WriteCloser, error) {...}}
cfg.EventHandler = handler
syncer := NewBinlogSyncer(cfg)
syncer.StartBackup(...)
```
and require changes to `StartBackup`/`StartBackupWithHandler`.

would you suggest doing that check elsewhere or changing the logic in some other way?

Reply:
I'll think about it tomorrow

Reply:
Currently I think the old `StartBackupWithHandler` is not very extendable with the new `EventHandler` interface. The old behaviour of `StartBackupWithHandler` includes switching to a new file for every `FORMAT_DESCRIPTION_EVENT`, and that is also a part of the functionality of `EventHandler` which can handle all types of events (I see the `BackupEventHandler` wraps on it and provides the full functionality). If we want to reuse the old `StartBackupWithHandler`, the code is not clear at these aspects:

- there are two `if b.cfg.SyncMode == SyncModeSync {...} else {...}`, one in `StartBackupWithHandler` and the other in `handleEventAndACK`. I think the duplication raises from the unclearness of the responsibility of the two functions.
- `SetEventHandler` + `StartSync` is more powerful than `StartBackupWithHandler`. For example, user can upload the binlog files to cloud storage or choose fsync strategy. But `StartBackupWithHandler` limits them to do so.
- The expected bahaviour is determined in two place, one in `BinlogSyncerConfig.SyncMode` the other in `handler` argument of `StartBackupWithHandler`. It's more clear to use a unified place.
- I think in no case `BinlogSyncer` should support change the `EventHandler` on the fly.

So I think you can leave alone the old `StartBackup` and `StartBackupWithHandler` (maybe add some checking logic to report error on new configurations), and write a new method like `(*BinlogSyncer) StartBackupWithEventHandler(...)` or wrap a new struct

```golang
type BinlogBackuper struct {
  s BinlogSyncer
}
```

to implement the new behaviour. And I prefer

```golang
type BinlogSyncerConfig struct {
...
  EventHandleMode EventHandleMode
  EventHandler EventHandler
}
```

or 

```golang
type BinlogSyncerConfig struct {
...
  SyncedEventHandler EventHandler // a non-nil handler will EventHandler.HandleEvent instead of GetEvent to handle event.
}
```

to initialize.

Reply:
feel free to discuss before start changing the code

Reply:
I updated this by replacing `SyncMode` and `EventHandleMode` with `SynchronousEventHandler`. I added `StartSynchronousBackup` and added an assertion to `StartBackupWithHandler` to check that `SynchronousEventHandler` is not set. I think this is a good compromise of simplicity that maintains backward compatibility while keeping clean interfaces. Let me know what you think and thanks again for your review.